types/encoding.go

```diff
@@ -0,0 +1,153 @@
+package types
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/NebulousLabs/Sia/encoding"
+)
+
+// MarshalSia implements the encoding.SiaMarshaler interface.
+func (t Transaction) MarshalSia(w io.Writer) error {
+	length := make([]byte, 8)
+	enc := encoding.NewEncoder(w)
+	binary.LittleEndian.PutUint64(length, uint64(len((t.SiacoinInputs))))
+	w.Write(length)
+	for i := range t.SiacoinInputs {
+		t.SiacoinInputs[i].MarshalSia(w)
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.SiacoinOutputs))))
+	w.Write(length)
+	for i := range t.SiacoinOutputs {
+		t.SiacoinOutputs[i].MarshalSia(w)
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.FileContracts))))
+	w.Write(length)
+	for i := range t.FileContracts {
+		enc.Encode(t.FileContracts[i])
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.FileContractRevisions))))
+	w.Write(length)
+	for i := range t.FileContractRevisions {
+		enc.Encode(t.FileContractRevisions[i])
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.StorageProofs))))
+	w.Write(length)
+	for i := range t.StorageProofs {
+		enc.Encode(t.StorageProofs[i])
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.SiafundInputs))))
+	w.Write(length)
+	for i := range t.SiafundInputs {
+		enc.Encode(t.SiafundInputs[i])
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.SiafundOutputs))))
+	w.Write(length)
+	for i := range t.SiafundOutputs {
+		t.SiafundOutputs[i].MarshalSia(w)
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.MinerFees))))
+	w.Write(length)
+	for i := range t.MinerFees {
+		t.MinerFees[i].MarshalSia(w)
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.ArbitraryData))))
+	w.Write(length)
+	for i := range t.ArbitraryData {
+		binary.LittleEndian.PutUint64(length, uint64(len((t.ArbitraryData[i]))))
+		w.Write(length)
+		w.Write(t.ArbitraryData[i])
+	}
+	binary.LittleEndian.PutUint64(length, uint64(len((t.TransactionSignatures))))
+	w.Write(length)
+	for i := range t.TransactionSignatures {
+		err := t.TransactionSignatures[i].MarshalSia(w)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
This is definitely a case where I would want to see some tests that prove equivalence for full types. It'd be relatively easy  to just miss one field here.